lib/internal/main/watch_mode.js

```diff
@@ -0,0 +1,125 @@
+const { inspect } = require('util');
+const { setTimeout, clearTimeout } = require('timers');
+const { resolve } = require('path');
+const { once, on } = require('events');
+
+
+prepareMainThreadExecution(false, false);
+markBootstrapComplete();
+
+// TODO(MoLow): Make kill signal configurable
+const kKillSignal = 'SIGTERM';
+const kShouldFilterModules = getOptionValue('--watch-path').length === 0;
+const kWatchedPaths = getOptionValue('--watch-path').length ?
+  ArrayPrototypeMap(getOptionValue('--watch-path'), (path) => resolve(path)) :
+  [];
+const kCommand = ArrayPrototypeSlice(process.argv, 1);
+const kCommandStr = inspect(ArrayPrototypeJoin(kCommand, ' '));
+const args = ArrayPrototypeFilter(process.execArgv, (x, i, arr) =>
+  x !== '--watch-path' && arr[i - 1] !== '--watch-path' && x !== '--watch');
+ArrayPrototypePush(args, '--watch-report-ipc');
+ArrayPrototypePushApply(args, kCommand);
+
+const watcher = new FilesWatcher({ throttle: 500, mode: kShouldFilterModules ? 'filter' : 'all' });
+kWatchedPaths.forEach((p) => watcher.watchPath(p));
+let graceTimer;
+let child;
+let exited;
+
+function start() {
+  // Spawning in detached mode so node can control when signals are forwarded
+  exited = false;
+  const stdio = kShouldFilterModules ? ['inherit', 'inherit', 'inherit', 'ipc'] : undefined;
+  child = spawn(process.execPath, args, { stdio, detached: true });
+  watcher.watchChildProcessModules(child);
+  child.once('exit', (code) => {
+    exited = true;
+    if (code === 0) {
+      process.stdout.write(`${blue}Completed running ${kCommandStr}${white}\n`);
+    } else {
+      process.stdout.write(`${red}Failed running ${kCommandStr}${white}\n`);
+    }
+  });
+}
+
+async function killAndWait(signal = kKillSignal) {
+  child?.removeAllListeners();
+  if (!child || child.killed || exited) {
+    return;
+  }
+  const onExit = once(child, 'exit');
+  child.kill(signal);
+  const { 0: exitCode } = await onExit;
+  return exitCode;
+}
+
+function reportGracefulTermination() {
+  // Log if process takes more than 500ms to stop
+  let reported = false;
+  clearTimeout(graceTimer);
+  graceTimer = setTimeout(() => {
+    reported = true;
+    process.stdout.write(`${blue}Waiting for graceful termination...${white}\n`);
+  }, 500).unref();
+  return () => {
+    clearTimeout(graceTimer);
+    if (reported) {
+      process.stdout.write(`${clear}${green}Gracefully restarted ${kCommandStr}${white}\n`);
+    }
+  };
+}
+
+async function stop() {
+  watcher.clearFileFilters();
+  const clearGraceReport = reportGracefulTermination();
+  await killAndWait();
+  clearGraceReport();
+}
+
+async function restart() {
+  process.stdout.write(`${clear}${green}Restarting ${kCommandStr}${white}\n`);
+  await stop();
+  start();
+}
+
+(async () => {
+  emitExperimentalWarning('Watch mode');
+  start();
+
+  // eslint-disable-next-line no-unused-vars
+  for await (const _ of on(watcher, 'changed')) {
+    await restart();
+  }
```

Comment:
I added added a message printed on stderr when this throws, is that the sort of handling you were thinking of?

Reply:
Hum I'm not sure that's better, crashing seems like a logical behavior if `restart()` rejects – I believe it would only reject if something very wrong happened, e.g. if `process.execPath` has been deleted or something.

Reply:
so you think `process.exit(1)` + printing the error should do?

Reply:
Yes, that would do. Please wrap the whole `for await` loop, as I find it stressful to have an `await` statement without any error handling 😅